File: backend/matching-service/app/controller/matching_controller.py
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, Header
from sqlalchemy.orm import Session
from typing import List
import httpx
import os

from app.database import get_db
from app.service.matching_service import MatchingService
from app.schemas.matching_schema import ConnectionRequest, ConnectionUpdate, ConnectionResponse, MatchSuggestion

router = APIRouter(prefix="/matching", tags=["matching"])

USER_SERVICE_URL = os.getenv("USER_SERVICE_URL", "http://user-service:8001")


def _user_id(x_user_id: str = Header(...)) -> UUID:
    try:
        return UUID(x_user_id)
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid user ID")


async def _get_user_profile(user_id: UUID) -> dict:
    async with httpx.AsyncClient() as client:
        resp = await client.get(f"{USER_SERVICE_URL}/api/v1/users/{user_id}", timeout=10.0)
        if resp.status_code == 200:
            return resp.json()
    return {}


async def _enrich_connections(connections: list, current_user_id: UUID) -> list:
    user_cache = {}

    async def fetch(uid_str):
        if uid_str not in user_cache:
            try:
                user_cache[uid_str] = await _get_user_profile(UUID(uid_str))
            except Exception:
                user_cache[uid_str] = {}
        return user_cache[uid_str]

    enriched = []
    for c in connections:
        await fetch(c["requester_id"])
        await fetch(c["addressee_id"])
        c["is_requester"] = c["requester_id"] == str(current_user_id)
        req = user_cache.get(c["requester_id"], {})
        adr = user_cache.get(c["addressee_id"], {})
        c["requester"] = {k: req.get(k) for k in ("id", "username", "first_name", "last_name", "fitness_level", "city")} if req else None
        c["addressee"] = {k: adr.get(k) for k in ("id", "username", "first_name", "last_name", "fitness_level", "city")} if adr else None
        enriched.append(c)
    return enriched


@router.get("/suggestions", response_model=List[MatchSuggestion])
async def get_suggestions(user_id: UUID = Depends(_user_id), db: Session = Depends(get_db)):
    profile = await _get_user_profile(user_id)
    if not profile:
        raise HTTPException(status_code=404, detail="User profile not found")
    service = MatchingService(db)
    results = await service.get_suggestions(user_id, profile)
    return [MatchSuggestion(**r) for r in results if r.get("id")]


@router.get("/browse", response_model=List[MatchSuggestion])
async def browse(user_id: UUID = Depends(_user_id), db: Session = Depends(get_db)):
    profile = await _get_user_profile(user_id)
    service = MatchingService(db)
    results = await service.get_suggestions(user_id, profile)
    return [MatchSuggestion(**r) for r in results if r.get("id")]


@router.get("/partners")
async def get_partners(user_id: UUID = Depends(_user_id), db: Session = Depends(get_db)):
    service = MatchingService(db)
    partner_ids = service.get_partners(user_id)
    partners = []
    for pid in partner_ids:
        p = await _get_user_profile(pid)
        if p:
            partners.append(p)
    return partners


@router.get("/connections", response_model=List[ConnectionResponse])
async def get_connections(user_id: UUID = Depends(_user_id), db: Session = Depends(get_db)):
    service = MatchingService(db)
    raw = service.get_connections(user_id)
    enriched = await _enrich_connections(raw, user_id)
    return [ConnectionResponse(**c) for c in enriched]


@router.post("/connections", response_model=ConnectionResponse, status_code=201)
def send_connection(
    data: ConnectionRequest,
    user_id: UUID = Depends(_user_id),
    db: Session = Depends(get_db)
):
    service = MatchingService(db)
    result = service.send_connection(user_id, UUID(data.addressee_id), data.match_score)
    if "error" in result:
        raise HTTPException(status_code=409, detail=result["error"])
    return ConnectionResponse(**result)


@router.put("/connections/{connection_id}", response_model=ConnectionResponse)
def update_connection(
    connection_id: UUID,
    data: ConnectionUpdate,
    user_id: UUID = Depends(_user_id),
    db: Session = Depends(get_db)
):
    service = MatchingService(db)
    result = service.update_connection(connection_id, user_id, data.status)
    if not result:
        raise HTTPException(status_code=404, detail="Connection not found")
    if "error" in result:
        raise HTTPException(status_code=403, detail=result["error"])
    return ConnectionResponse(**result)


@router.delete("/connections/{connection_id}", status_code=204)
def delete_connection(
    connection_id: UUID,
    user_id: UUID = Depends(_user_id),
    db: Session = Depends(get_db)
):
    service = MatchingService(db)
    if not service.delete_connection(connection_id, user_id):
        raise HTTPException(status_code=404, detail="Connection not found or access denied")

File: backend/matching-service/src/matching/matching.controller.ts
import {
  Controller, Get, Post, Put, Delete, Body, Param, Headers, Query,
  HttpCode, NotFoundException, BadRequestException, ConflictException, ForbiddenException,
} from '@nestjs/common';
import { MatchingService } from './matching.service';

@Controller('matching')
export class MatchingController {
  constructor(private service: MatchingService) {}

  private userId(id: string) {
    if (!id) throw new BadRequestException('Missing x-user-id header');
    return id;
  }

  @Get('health')
  health() { return { status: 'healthy', service: 'matching-service' }; }

  @Get('suggestions')
  async suggestions(@Headers('x-user-id') uid: string) {
    const userId = this.userId(uid);
    const profile = await this.service.fetchUser(userId);
    if (!profile) throw new NotFoundException('User profile not found');
    return this.service.getSuggestions(userId, profile);
  }

  @Get('browse')
  async browse(@Headers('x-user-id') uid: string, @Query() query: Record<string, string>) {
    const userId = this.userId(uid);
    const profile = await this.service.fetchUser(userId);
    const { fitness_level, primary_goal, city } = query;
    const filters: Record<string, string> = {};
    if (fitness_level) filters.fitness_level = fitness_level;
    if (primary_goal) filters.primary_goal = primary_goal;
    if (city) filters.city = city;
    return this.service.getSuggestions(userId, profile || {}, filters);
  }

  @Get('partners')
  async partners(@Headers('x-user-id') uid: string) {
    const userId = this.userId(uid);
    const ids = await this.service.getPartnerIds(userId);
    return Promise.all(ids.map((id) => this.service.fetchUser(id))).then((r) => r.filter(Boolean));
  }

  @Get('connections')
  async connections(@Headers('x-user-id') uid: string) {
    const userId = this.userId(uid);
    const raw = await this.service.getConnections(userId);
    return this.service.enrichConnections(raw, userId);
  }

  @Post('connections')
  @HttpCode(201)
  async sendRequest(@Headers('x-user-id') uid: string, @Body() body: any) {
    const userId = this.userId(uid);
    const result = await this.service.sendConnection(userId, body.addressee_id, body.match_score);
    if ('error' in result) throw new ConflictException(result.error);
    return result;
  }

  @Put('connections/:id')
  async updateConn(@Param('id') id: string, @Headers('x-user-id') uid: string, @Body() body: any) {
    const userId = this.userId(uid);
    const result = await this.service.updateConnection(id, userId, body.status);
    if (!result) throw new NotFoundException('Connection not found');
    if ('error' in result) throw new ForbiddenException(result.error);
    return result;
  }

  @Delete('connections/:id')
  @HttpCode(204)
  async deleteConn(@Param('id') id: string, @Headers('x-user-id') uid: string) {
    const userId = this.userId(uid);
    const ok = await this.service.deleteConnection(id, userId);
    if (!ok) throw new NotFoundException('Connection not found');
  }
}
